### PR TITLE
chore: add use_gpu for cifar finetuning

### PR DESCRIPTION
**Pull request by kcelia**

### CODE REVIEW

**Review by andrei-stoian-zama — CHANGES_REQUESTED** [review]
Thanks, but you need to use the new CML GPU api instead of the Concrete one

File: use_case_examples/cifar/cifar_brevitas_finetuning/FromImageNetToCifar.ipynb [hunk 1]
```diff
@@ -64,7 +65,10 @@
    "source": [
     "dataset_name = \"CIFAR_100\"\n",
     "\n",
-    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
+    "# To use the CUDA backend for the Concrete compiler, install the GPU-supported version.\n",
+    "# Refer to the official guide:\n",
+    "# https://github.com/zama-ai/concrete-ml/blob/main/docs/guides/using_gpu.md#usage-preqreuisites\n",
```
Comment (andrei-stoian-zama) [inline use_case_examples/cifar/cifar_brevitas_finetuning/FromImageNetToCifar.ipynb:70]: we'll need to merge the other PR so that the docs are published, then link to the actual doc

File: use_case_examples/cifar/cifar_brevitas_finetuning/CifarQuantizationAwareTraining.ipynb [hunk 2]
```diff
@@ -88,7 +89,10 @@
    "source": [
     "bit = 5\n",
     "\n",
-    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
+    "# To use the CUDA backend for the Concrete compiler, install the GPU-supported version.\n",
+    "# Refer to the official guide:\n",
+    "# https://github.com/zama-ai/concrete-ml/blob/main/docs/guides/using_gpu.md\n",
```
Comment (andrei-stoian-zama) [inline use_case_examples/cifar/cifar_brevitas_finetuning/CifarQuantizationAwareTraining.ipynb:94]: same, we need a link to the actual doc on docs.zama.ai

File: use_case_examples/cifar/cifar_brevitas_finetuning/CifarInFheWithSmallerAccumulators.ipynb [hunk 3]
```diff
@@ -59,7 +61,10 @@
     "seed = 42\n",
     "rounding_thresholds_bits = [8, 7, 6, 5, 3]\n",
     "\n",
-    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\""
+    "# To use the CUDA backend for the Concrete compiler, install the GPU-supported version.\n",
+    "# Refer to the official guide:\n",
+    "# https://github.com/zama-ai/concrete-ml/blob/main/docs/guides/using_gpu.md\n",
```
Comment (andrei-stoian-zama) [inline use_case_examples/cifar/cifar_brevitas_finetuning/CifarInFheWithSmallerAccumulators.ipynb:66]: need the actual link here too

File: use_case_examples/cifar/cifar_brevitas_finetuning/CifarInFhe.ipynb [hunk 4]
```diff
@@ -62,7 +63,10 @@
     "bit = 5\n",
     "seed = 42\n",
     "\n",
-    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
+    "# To use the CUDA backend for the Concrete compiler, install the GPU-supported version.\n",
+    "# Refer to the official guide:\n",
+    "# https://github.com/zama-ai/concrete-ml/blob/main/docs/guides/using_gpu.md\n",
+    "device = \"cuda\" if check_gpu_available() else \"cpu\"\n",
```
Comment (andrei-stoian-zama) [inline use_case_examples/cifar/cifar_brevitas_finetuning/CifarInFhe.ipynb:69]: the PyTorch `device` and CML `device` may differ. You should use the original

`device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",` 

for pytorch training, and a separate device variable for CML compilation (based on the CML check for gpu availability)

**Review by andrei-stoian-zama — CHANGES_REQUESTED** [review]
@kcelia Can you please confirm you can run these notebooks on your machine ?

File: use_case_examples/cifar/cifar_brevitas_finetuning/FromImageNetToCifar.ipynb (hunk 1 above)
Comment (kcelia) [reply use_case_examples/cifar/cifar_brevitas_finetuning/FromImageNetToCifar.ipynb:70]: The PR is ready

File: use_case_examples/resnet/run_resnet18_fhe.py [hunk 5]
```diff
@@ -53,7 +54,12 @@ def evaluate_model(model, processor):
 
 
 def compile_model(
-    model, images, n_bits, rounding_threshold_bits=None, fhe_mode="disable", use_gpu=False
+    model,
+    images,
+    n_bits,
+    rounding_threshold_bits=None,
+    fhe_mode="disable",
+    compilation_device="cpu",
```
Comment (jfrery) [inline use_case_examples/resnet/run_resnet18_fhe.py:62]: I preferred `use_gpu` or just `device=...` to do the same as torch.

Comment (kcelia) [reply use_case_examples/resnet/run_resnet18_fhe.py:62]: "@jfrery, `use_device` might sound like a boolean type, but in the compilation functions, it's actually a string that can be either 'cpu' or 'gpu'. 

The reason it's not simply named `device` is because the PyTorch device and the CML one may differ. This distinction helps differentiate between the device used for PyTorch training/evaluation and the one used for CML compilation.

Comment (andrei-stoian-zama) [reply use_case_examples/resnet/run_resnet18_fhe.py:62]: @jfrery the API was decided some time ago:

- https://zama-ai.slack.com/archives/C0455QDGT6C/p1723040023589979
- https://zama-ai.slack.com/archives/C0455QDGT6C/p1724860575912049

File: use_case_examples/cifar/cifar_brevitas_training/evaluate_torch_cml.py [hunk 6]
```diff
@@ -123,6 +123,7 @@ def main(args):
                 if rounding_threshold_bits is not None
                 else None
             ),
+            device=COMPILATION_DEVICE,
```
Comment (andrei-stoian-zama) [inline use_case_examples/cifar/cifar_brevitas_training/evaluate_torch_cml.py:126]: this isn't defined anywhere

File: use_case_examples/cifar/cifar_brevitas_finetuning/CifarQuantizationAwareTraining.ipynb [hunk 7]
```diff
@@ -87,12 +88,32 @@
    ],
    "source": [
     "bit = 5\n",
+    "seed = 42\n",
```
Comment (andrei-stoian-zama) [inline use_case_examples/cifar/cifar_brevitas_finetuning/CifarQuantizationAwareTraining.ipynb:91]: why was this added ?

**Review by andrei-stoian-zama — CHANGES_REQUESTED** [review]
one small error remains

File: use_case_examples/cifar/cifar_brevitas_training/evaluate_torch_cml.py (hunk 6 above)
Comment (kcelia) [reply use_case_examples/cifar/cifar_brevitas_training/evaluate_torch_cml.py]: I corrected it, when I tested on my machine

File: use_case_examples/cifar/cifar_brevitas_finetuning/CifarQuantizationAwareTraining.ipynb (hunk 7 above)
Comment (kcelia) [reply use_case_examples/cifar/cifar_brevitas_finetuning/CifarQuantizationAwareTraining.ipynb:91]: Bad copy past

**Comment by kcelia** [comment]
I tested on my machine, all the scripts are working, except,

File: use_case_examples/resnet/run_resnet18_fhe.py (hunk 5 above)
Comment (jfrery) [reply use_case_examples/resnet/run_resnet18_fhe.py:62]: `device="cuda|cpu"` is perfect yes but here we have compilation_device. 

> The reason it's not simply named device is because the PyTorch device and the CML one may differ.

Well as far as I know we try to stay close to the torch API. We did it for sklearn and the others. I don't see why now we would want to differ. device seems the right choice here.

Comment (kcelia) [reply use_case_examples/resnet/run_resnet18_fhe.py:62]: We assume that the compile_device may differ from the device used to train the model.

Additionally, on standard machines, setting compile_device="cpu" is currently faster than compile_device="cuda". So, it's better to keep them separate.

**Review by andrei-stoian-zama — APPROVED** [review]
